WordPress/gutenberg
# Global Styles: Allow references to values in other locations in the tree

## What?
There are many times when a block would like to access properties from the global styles context but is unable to do so. Sometimes that isn't a problem as the CSS cascade takes care of it for us, but in the case where blocks want to apply a setting to a different property, there's no mechanism to do so. This PR proposes that we add a new convention to theme.json which allows one property to set itself to the value of another, for example buttons could do this:

```
{
	"color": {
		"text": { ref: "styles.color.background" }
	}
}
```

Global Styles will convert `{ ref: "styles.color.background" }` to whatever is the value at `styles > color > background` in theme.json.

## Why?
This will allow blocks to be a lot less opinionated in their styles and simply inherit from theme settings.

## How?
The main part of this change is in `get_property_value`, which adds some lines of code to detect whether the value starts with `styles` or `settings`. If it does then we replace it with the result of _wp_array_get. The rest of the changes are necessary to pass $theme_json down to this function, since it's static. This was first attempted in https://github.com/WordPress/gutenberg/pull/39432.

## Testing Instructions
- Switch to emptytheme
- Add these blocks to a post: Button, File, Search
- Check that the colors of these button are the inverse of the colors of the background and text of the site.

## Screenshots or screencast 
<img width="528" alt="Screenshot 2022-06-13 at 17 19 37" src="https://user-images.githubusercontent.com/275961/173399179-54adcb5e-3f09-4a13-a03b-02379c0ac30f.png">


## Note
This change has a few side effects that we'll need to be careful with
1. The default colors for theme.json are now fixed, so they won't fall back to the browser defaults.
2. Search block buttons will now match other buttons in visual appearance, which they didn't before.

cc @WordPress/block-themers
